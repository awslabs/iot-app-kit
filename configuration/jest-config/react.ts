export default {
  setupFilesAfterEnv: ['jest-extended/all'],
  maxWorkers: '50%',
  moduleFileExtensions: ['js', 'ts', 'jsx', 'tsx'],
  collectCoverage: true,
  testEnvironment: 'jsdom',
  transform: {
    '.*\\.(tsx?|jsx?)$': [
      '@swc/jest',
      {
        jsc: {
          transform: {
            react: {
              runtime: 'automatic',
            },
          },
        },
      },
    ],
  },
  transformIgnorePatterns: [],
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy',
  },
};
