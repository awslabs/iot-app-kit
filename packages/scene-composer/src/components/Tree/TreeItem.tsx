import { Button } from '@awsui/components-react';
import React, { ComponentPropsWithRef, FC, ReactNode, useCallback } from 'react';
import { useIntl } from 'react-intl';

import RadioButton from '../RadioButton';
import { useStore } from '../../store';
import { useSceneComposerId } from '../../common/sceneComposerIdContext';

export type SelectionMode = 'single' | 'multi';

interface TreeItemInnerProps {
  label?: string;
  selected?: boolean;
  selectable?: boolean;
  className?: string;
  onActivated?(): Promise<void> | void;
  onSelected?(newState: boolean, e?: any): Promise<void> | void;
}

export interface TreeItemProps extends TreeItemInnerProps, ComponentPropsWithRef<'li'> {
  labelNode: ReactNode;
  labelText: string;
  selectionMode?: SelectionMode;

  // Expandable
  expandable?: boolean;
  expanded?: boolean;
  onExpand?<TEvent>(newState: boolean, e: TEvent): void | Promise<void>;
}

const TreeItemInner: FC<TreeItemInnerProps & { children: ReactNode }> = ({
  children,
  label,
  onSelected = /* istanbul ignore next */ () => {},
  selectable = true,
  selected = false,
  className = '',
  onActivated = () => {},
}) => {
  const sceneComposerId = useSceneComposerId();
  const setSelectedSceneNodeRef = useStore(sceneComposerId).getState().setSelectedSceneNodeRef;

  const selectFromChildren = () => {
    (children as any)?.forEach((child) => {
      return child ? setSelectedSceneNodeRef(child.props?.objectRef) : null;
    });
  };

  const toggle = useCallback(
    (e) => {
      if (selected) setSelectedSceneNodeRef(undefined);
      return (children) =>
        children.every((child) => child === true) ? selectFromChildren() : onSelected(!selected, e);
    },
    [selected, onSelected],
  );

  return (
    <label
      className={`tm-tree-item-inner ${selected ? 'selected' : ''} ${className}`.trimEnd()}
      onClick={toggle}
      onDoubleClick={onActivated}
      aria-selected={selected as boolean}
    >
      {selectable && (
        <>
          <RadioButton selected={selected} toggle={onSelected} label={label} />
          {children}
        </>
      )}
      {!selectable && children}
    </label>
  );
};

TreeItemInner.displayName = 'TreeItemInner';

const TreeItem = React.forwardRef<HTMLLIElement, TreeItemProps>(
  (
    {
      className = '',
      labelNode,
      labelText,
      children,
      selected,
      selectable,
      onActivated,
      onSelected,
      selectionMode = 'single',
      expandable,
      expanded,
      onExpand = () => {},
      ...props
    }: TreeItemProps,
    ref,
  ) => {
    const { formatMessage } = useIntl();
    const expandHandler = useCallback(
      async (e) => {
        e.stopPropagation(); // Prevent bubbling that would result in triggering selected.
        await onExpand(!expanded, e);
      },
      [onExpand, expanded],
    );

    return (
      <li
        ref={ref}
        className={`tm-tree-item${expandable ? ' expandable' : ''} ${className}`.trim()}
        role='treeitem'
        {...props}
      >
        <TreeItemInner
          selected={selected}
          selectable={selectable}
          onActivated={onActivated}
          onSelected={onSelected}
          label={labelText}
        >
          {expandable && (
            <Button
              className='tm-tree-item-expand-btn'
              variant='inline-icon'
              onClick={expandHandler}
              iconName={`treeview-${expanded ? 'collapse' : 'expand'}`}
              ariaExpanded={expanded}
              ariaLabel={`${
                expanded
                  ? formatMessage({
                      defaultMessage: 'Collapse',
                      description: 'Label for button that triggers a collapsable section',
                    })
                  : formatMessage({
                      defaultMessage: 'Expand',
                      description: 'Label for button that triggers a expandable section',
                    })
              } ${labelText}`}
            />
          )}
          {labelNode}
        </TreeItemInner>
        {!expandable || (expandable && expanded) ? children : null}
      </li>
    );
  },
);

TreeItem.displayName = 'TreeItem';

export default TreeItem;
