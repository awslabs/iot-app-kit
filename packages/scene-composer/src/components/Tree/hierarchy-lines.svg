<svg viewBox="0 0 2 4" preserveAspectRatio="none" style="margin: -0.5rem 0px 0px; padding-left: 0.3rem; position: absolute; inset: 0px 2rem 0px 0rem; width: 2rem; height: 4rem;">
  <line x1="50%" x2="100%" y1="50%" y2="50%" style="stroke: var(--color-border-divider-default-k96bks, #eaeded); stroke-width: 2; vector-effect: non-scaling-stroke;"></line>
  <line x1="50%" x2="50%" y1="0%" y2="50%" style="stroke: var(--color-border-divider-default-k96bks, #eaeded); stroke-width: 2; vector-effect: non-scaling-stroke;"></line>
  <line x1="100%" x2="50%" y1="50%" y2="50%" style="stroke: var(--color-border-divider-default-k96bks, #eaeded); stroke-width: 2; vector-effect: non-scaling-stroke;"></line>
  <line x1="0%" x2="50%" y1="50%" y2="50%" style="stroke: var(--color-border-divider-default-k96bks, #eaeded); stroke-width: 2; vector-effect: non-scaling-stroke;"></line>
</svg>
