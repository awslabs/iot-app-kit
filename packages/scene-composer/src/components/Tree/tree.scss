$background: #2a2e33;
$background-gradient: linear-gradient(90deg, rgba(0, 0, 0, 0) 0%, rgba(42, 46, 51, 1) 38%, rgba(42, 46, 51, 1) 100%);
$background-hover: #414750;
$border: 1px solid $background;
$text-color: #d5dbdb;
$active-highlight: #00a1c9;
$active-bg: #12293b;
$active-bg-gradient: linear-gradient(90deg, rgba(0, 0, 0, 0) 0%, rgba(18, 41, 59, 1) 38%, rgba(18, 41, 59, 1) 100%);
$indent: 1rem;
$active-color: #ec7211;

@mixin selected {
  border-top: 1px solid $active-highlight;
  border-bottom: 1px solid $active-highlight;
  background: $active-bg;
}
@mixin selected-action {
  @include selected;
  background: $active-bg-gradient;
  margin-top: -1px;
  margin-bottom: -1px;
}
@mixin hover {
  @include selected;
  color: $active-color;
}

#tm-main {
  margin-bottom: 5px;

  // Adjustment to ensure we're able to keep layouts inline with Polaris modifications
  p {
    line-height: 2rem;
  }

  .tm-tree {
    border-top: $border;
    list-style-position: outside;
    list-style-type: none;
    margin: 0;
    padding: 0;
    position: relative;
    user-select: none;

    &.drop,
    .drop {
      border-bottom: 3px solid #fff;
    }

    .tm-tree-item {
      padding-left: $indent * 2;
      padding-top: 0;

      .tm-draggable[draggable=true]::before {
        margin-top: 1.6rem;
    }

      .tm-tree-item-inner {
        cursor: pointer;
        margin-left: -2rem;
        padding-left: 2rem;
        align-items: center;
        display: flex;

        .tm-tree-item-expand-btn {
          min-width: unset;
        }

        .tm-scene-node-label {
          border-top: 1px solid $background-hover;
          display: grid;
          grid-template-columns: repeat(20, 1fr);
          align-items: center;

          .tm-scene-node-label-inner {
            flex: 0 0 100%;
            padding: 13px 50px 14px 20px;
            width: 100vw;
            margin: 0;
            white-space: nowrap;
          }

          .actions {
            background-color: var(--color-background-container-content-o88wxm, #ffffff); // fallback for gradient
            background: $background-gradient;
            position: sticky;
            right: 0;
            padding-right: 12px;
            color: $active-color;
          }
        }

        &:hover {
          margin-top: -1px; // Cancells out movement from hover/selected border
          margin-bottom: 0px; // Cancells out movement from hover/selected border
          @include hover;

          .tm-scene-node-label-inner {
            border: none;
            @include selected;
            margin-bottom: -1px;
            margin-top: -2px;
          }

          .actions {
            @include selected-action;
            margin-top: -3px;
            margin-bottom: -1px;
            padding: 12px 12px 13px 50px;
          }

          .tm-visibility-toggle,
          .tm-tree-item-expand-btn {
            svg {
              color: $active-color;
            }
          }
        }

        &.selected {
          @include selected;

          .tm-scene-node-label-inner {
            border: none;
            @include selected;
            margin-bottom: -1px;
            margin-top: -2px;
            margin-right: -30px;
          }

          .actions {
            @include selected-action;
            margin-top: -3px;
            margin-bottom: -1px;
            padding-bottom: 13px;
            padding: 12px 12px 13px 50px;
          }
        }
      }
    }

    // nested tree
    .tm-tree {
      border-top: none;
      position: initial;
      padding-left: 0;
      padding-bottom: 0;
    }
  }
}
