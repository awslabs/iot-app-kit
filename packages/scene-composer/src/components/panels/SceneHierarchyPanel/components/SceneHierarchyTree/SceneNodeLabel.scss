.tm-scene-node-label {
  .actions {
    float: right;
    svg {
      cursor: pointer;
    }
  }
}

.error {
  color: red;
}
