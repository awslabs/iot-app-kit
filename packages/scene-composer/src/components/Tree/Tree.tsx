import React from 'react';

import './tree.scss';

export interface TreeProps extends React.ComponentPropsWithoutRef<'ol'> {}

const Tree = React.forwardRef<HTMLOListElement, TreeProps>(({ children, className = '', ...props }: TreeProps, ref) => {
  return (
    <ol ref={ref} role='tree' className={`tm-tree ${className}`.trim()} {...props}>
      {children}
    </ol>
  );
});

Tree.displayName = 'Tree';

export default Tree;
