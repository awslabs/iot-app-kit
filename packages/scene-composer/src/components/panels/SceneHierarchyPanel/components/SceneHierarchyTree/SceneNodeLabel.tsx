import React, { FC, useCallback, useState } from 'react';
import { Button, Icon } from '@awsui/components-react';

import VisibilityToggle from '../../../../../components/VisibilityToggle';
import { KnownComponentType } from '../../../../../interfaces';
import './SceneNodeLabel.scss';
import { DeleteSvg } from '../../../../../assets/svgs';
import { useSceneHierarchyData } from '../../SceneHierarchyDataProvider';

import ComponentTypeIcon from './ComponentTypeIcon';

export interface SceneNodeLabelProps {
  objectRef: string;
  labelText: string;
  componentTypes?: string[];
}

const SceneNodeLabel: FC<SceneNodeLabelProps> = ({ objectRef, labelText, componentTypes }) => {
  const { show, hide, remove, validationErrors } = useSceneHierarchyData();
  const [visible, setVisible] = useState(true);

  const error = validationErrors[objectRef];

  const componentTypeIcons = componentTypes
    ?.filter((type) => !!type && Object.keys(KnownComponentType).includes(type))
    .map((type) => <ComponentTypeIcon key={type} type={type} />);

  const toggleVisibility = useCallback(
    (newVisibility) => {
      if (newVisibility) {
        show(objectRef);
      } else {
        hide(objectRef);
      }

      setVisible(newVisibility);
    },
    [objectRef, visible, show, hide],
  );

  const onDelete = useCallback(() => {
    remove(objectRef);
  }, [objectRef]);

  return (
    <span className={`tm-scene-node-label ${error ? 'error' : ''}`.trim()} title={error}>
      {componentTypeIcons}
      <p className='tm-scene-node-label-inner'>{labelText}</p>
      <span className='actions'>
        {!!error && <Button onClick={onDelete} variant='inline-icon' iconSvg={DeleteSvg} />}
        <VisibilityToggle visible={visible} onToggle={toggleVisibility} />
      </span>
    </span>
  );
};

SceneNodeLabel.displayName = 'SceneNodeLabel';

export default SceneNodeLabel;
