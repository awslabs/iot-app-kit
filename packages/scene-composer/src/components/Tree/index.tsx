import Tree from './Tree';

export { default as TreeItem } from './TreeItem';

export default Tree;
