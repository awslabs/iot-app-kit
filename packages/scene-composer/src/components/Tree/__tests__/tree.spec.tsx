import React, { useRef } from 'react';
import { render } from '@testing-library/react';

import Tree, { TreeItem } from '..';
import SceneNodeLabel from '../../panels/SceneHierarchyPanel/components/SceneHierarchyTree/SceneNodeLabel';

jest.mock('react-dnd', () => ({
  useDrag: jest.fn(() => [1, useRef(2)]),
  useDrop: jest.fn(() => [1, useRef(2)]),
}));

jest.mock('../../panels/SceneHierarchyPanel/components/SceneHierarchyTree/SceneNodeLabel', () => (props) => (
  <div data-mocked='SceneNodeLabel'>{JSON.stringify(props)}</div>
));

const mockSceneNodeLabel = ({ label }) => <SceneNodeLabel objectRef='' labelText={label} componentTypes={['']} />;

describe('<Tree />', () => {
  it('should render appropriate structure', () => {
    const { container } = render(
      <Tree>
        <TreeItem labelNode={mockSceneNodeLabel({ label: 'Level 1' })} labelText='Level 1'>
          <Tree>
            <TreeItem labelNode={mockSceneNodeLabel({ label: 'Level 2' })} labelText='Level 2'>
              <Tree>
                <TreeItem labelNode={mockSceneNodeLabel({ label: 'Level 3' })} labelText='Level 3' />
              </Tree>
            </TreeItem>
          </Tree>
        </TreeItem>
        <TreeItem labelNode={mockSceneNodeLabel({ label: 'Level 1 Sibling' })} labelText='Level 1 Sibling' />
      </Tree>,
    );

    expect(container).toMatchSnapshot();
  });
});
