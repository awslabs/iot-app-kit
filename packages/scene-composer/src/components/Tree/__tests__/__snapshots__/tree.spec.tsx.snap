// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tree /> should render appropriate structure 1`] = `
<div>
  <ol
    class="tm-tree"
    role="tree"
  >
    <li
      class="tm-tree-item"
      role="treeitem"
    >
      <label
        aria-selected="false"
        class="tm-tree-item-inner"
      >
        <input
          type="radio"
          value="Level 1"
        />
        <div
          data-mocked="SceneNodeLabel"
        >
          {"objectRef":"","labelText":"Level 1","componentTypes":[""]}
        </div>
      </label>
      <ol
        class="tm-tree"
        role="tree"
      >
        <li
          class="tm-tree-item"
          role="treeitem"
        >
          <label
            aria-selected="false"
            class="tm-tree-item-inner"
          >
            <input
              type="radio"
              value="Level 2"
            />
            <div
              data-mocked="SceneNodeLabel"
            >
              {"objectRef":"","labelText":"Level 2","componentTypes":[""]}
            </div>
          </label>
          <ol
            class="tm-tree"
            role="tree"
          >
            <li
              class="tm-tree-item"
              role="treeitem"
            >
              <label
                aria-selected="false"
                class="tm-tree-item-inner"
              >
                <input
                  type="radio"
                  value="Level 3"
                />
                <div
                  data-mocked="SceneNodeLabel"
                >
                  {"objectRef":"","labelText":"Level 3","componentTypes":[""]}
                </div>
              </label>
            </li>
          </ol>
        </li>
      </ol>
    </li>
    <li
      class="tm-tree-item"
      role="treeitem"
    >
      <label
        aria-selected="false"
        class="tm-tree-item-inner"
      >
        <input
          type="radio"
          value="Level 1 Sibling"
        />
        <div
          data-mocked="SceneNodeLabel"
        >
          {"objectRef":"","labelText":"Level 1 Sibling","componentTypes":[""]}
        </div>
      </label>
    </li>
  </ol>
</div>
`;
